File: src/backend/mt/impl/frame.m
/*
 * Copyright (C) 2020 Recep Aslantas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../common.h"

GPU_HIDE
GPUFrame*
mt_beginFrame(GPUApi       *__restrict api,
              GPUSwapchain *__restrict swapchain) {
  GPUFrame           *frame;
  GPUTexture         *target;
  GPUTextureView     *targetView;
  GPUSwapchainMetal  *swapchainMtl;
  id<CAMetalDrawable> drawable;

  swapchainMtl = swapchain->_priv;
  if (!swapchainMtl || swapchainMtl->frameActive) {
    return NULL;
  }
  drawable     = [swapchainMtl->layer nextDrawable];
  if (!drawable) {
    return NULL;
  }

  [drawable retain];
  frame = &swapchainMtl->frame;
  target = &swapchainMtl->target;
  targetView = &swapchainMtl->targetView;
  memset(frame, 0, sizeof(*frame));
  memset(target, 0, sizeof(*target));
  memset(targetView, 0, sizeof(*targetView));

  target->_priv = drawable.texture;
  target->format = mt_formatFromNative(drawable.texture.pixelFormat);
  target->dimension = GPU_TEXTURE_DIMENSION_2D;
  target->width = (uint32_t)drawable.texture.width;
  target->height = (uint32_t)drawable.texture.height;
  target->depthOrLayers = 1;
  target->mipLevelCount = 1;
  target->sampleCount = 1;
  target->usage = GPU_TEXTURE_USAGE_COLOR_TARGET;
  target->_ownsNative = false;

  targetView->_priv = drawable.texture;
  targetView->_texture = target;
  targetView->format = target->format;
  targetView->viewType = GPU_TEXTURE_VIEW_2D;
  targetView->baseMipLevel = 0;
  targetView->mipLevelCount = 1;
  targetView->baseArrayLayer = 0;
  targetView->arrayLayerCount = 1;
  targetView->_ownsNative = false;

  frame->_priv      = swapchainMtl;
  frame->target     = target;
  frame->targetView = targetView;
  frame->drawable   = drawable;
  swapchainMtl->frameActive = true;

  return frame;
}

GPU_HIDE
void
mt_endFrame(GPUApi   *__restrict api,
            GPUFrame *__restrict frame) {
  GPUSwapchainMetal *swapchainMtl;

  (void)api;

  if (!frame)
    return;

  swapchainMtl = frame->_priv;
  [(id<CAMetalDrawable>)frame->drawable release];
  frame->drawable = NULL;
  if (swapchainMtl) {
    swapchainMtl->frameActive = false;
  }
}

GPU_HIDE
void
mt_initFrame(GPUApiFrame *api) {
  api->beginFrame = mt_beginFrame;
  api->endFrame   = mt_endFrame;
}

File: src/backend/dx12/impl/frame.c
/*
 * Copyright (C) 2020 Recep Aslantas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../common.h"
#include "../impl.h"

GPU_HIDE
GPUFrame*
dx12_beginFrame(GPUApi       * __restrict api,
                GPUSwapChain * __restrict swapChain) {
  GPUSwapChainDX12 *native;
  GPUFrameDX12     *frame;
  UINT              frameIndex;

  GPU__UNUSED(api);

  native = swapChain ? swapChain->_priv : NULL;
  if (!native || !native->swapChain || !native->frames ||
      native->imageCount == 0u || native->frameActive) {
    return NULL;
  }

  frameIndex = native->swapChain->lpVtbl->GetCurrentBackBufferIndex(
    native->swapChain
  );
  if (frameIndex >= native->imageCount) {
    return NULL;
  }

  frame = &native->frames[frameIndex];
  if (!dx12_waitQueueFence(native->queue,
                           frame->fenceValue,
                           native->frameEvent)) {
    return NULL;
  }

  native->frameIndex     = frameIndex;
  native->frameActive    = true;
  native->frameScheduled = false;
  frame->frame._priv      = native;
  frame->frame.target     = &frame->target;
  frame->frame.targetView = &frame->targetView;
  frame->frame.drawable   = frame;
  return &frame->frame;
}

GPU_HIDE
void
dx12_endFrame(GPUApi   * __restrict api,
              GPUFrame * __restrict frame) {
  GPUSwapChainDX12 *native;

  GPU__UNUSED(api);

  native = frame ? frame->_priv : NULL;
  if (!native || !native->frameActive) {
    return;
  }

  frame->drawable        = NULL;
  native->frameActive    = false;
  native->frameScheduled = false;
}

GPU_HIDE
void
dx12_schedulePresent(GPUCommandBuffer *cmdb, GPUFrame *frame) {
  GPUCommandBufferDX12 *command;
  GPUSwapChainDX12     *swapchain;

  command   = cmdb ? cmdb->_priv : NULL;
  swapchain = frame ? frame->_priv : NULL;
  if (!command || !swapchain || !frame->drawable ||
      !swapchain->frameActive || swapchain->frameScheduled ||
      command->presentSwapchain || command->owner != swapchain->queue) {
    return;
  }

  command->presentSwapchain  = swapchain;
  swapchain->frameScheduled  = true;
}

GPU_HIDE
void
dx12_initFrame(GPUApiFrame *api) {
  api->beginFrame = dx12_beginFrame;
  api->endFrame   = dx12_endFrame;
}

GPU_HIDE
void
dx12_initCmdbuf(GPUApiCommandBuffer *api) {
  api->presentDrawable = dx12_schedulePresent;
}
